File: benchmarks/static/corpus/go/taint.go
// M23 Phase 2: interprocedural taint for Go. A source-like parameter that
// reaches a command sink (exec.Command) is GF-304 (proven flow), distinct from
// the GF-404 shell-literal heuristic. These use exec.Command WITHOUT a shell
// literal so only the taint engine fires (no GF-404 overlap), isolating GF-304.
package p

import (
	"log"
	"os/exec"
)

// Direct: a tainted source parameter reaches the sink in the same function.
func run(userInput string) {
	exec.Command(userInput) // EXPECT GF-304
}

// Interprocedural: a source flows through a project-local call to the sink.
func dispatch(userQuery string) {
	forward(userQuery)
}

func forward(a string) {
	exec.Command(a) // EXPECT GF-304
}

// Sanitized before the sink — taint is cleared, so nothing must fire here.
func clean(userPath string) {
	v := sanitize(userPath)
	exec.Command(v)
}

func logUser(userInput string) {
	log.Printf("%s", userInput) // EXPECT GF-544
	log.Print("fixed")
}

File: benchmarks/static/corpus/rust/taint.rs
// SPDX-License-Identifier: Apache-2.0
// M23 Phase 2: interprocedural taint for Rust. A source-like parameter reaching
// a command sink (std::process::Command) is GF-304 (proven flow). These use
// Command::new(arg) WITHOUT a shell literal so only the taint engine fires (no
// GF-404 overlap), isolating GF-304.
fn run(user_input: &str) {
    std::process::Command::new(user_input); // EXPECT GF-304
}

fn dispatch(user_query: String) {
    forward(user_query);
}

fn forward(a: String) {
    std::process::Command::new(a); // EXPECT GF-304
}

fn clean(user_path: &str) {
    let v = sanitize(user_path);
    std::process::Command::new(v);
}

fn log_user(user_input: &str) {
    log::warn!("{}", user_input); // EXPECT GF-544
    log::info!("fixed");
}
